### PR TITLE
Add SparsePrivateUse1 in backend_to_string, layout_from_backend and check_base_legacy_new.

### PR DESCRIPTION
**Pull request by Shan19900305**
1) Using items stored in torch._tensor_classes to check item passed from python side;
2) Add SparsePrivateUse1 in backend_to_string, layout_from_backend and check_base_legacy_new;
3) Using more general API to get python module name in get_storage_obj and get_name functions.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @soulitzer @bdhirsh @ezyang @albanD 

Fixes #ISSUE_NUMBER

### CODE REVIEW

**Comment by Shan19900305** [comment]
@ezyang After more radical modification is not accept and i pushed a new pr, which is just make some APIs more general for all other devices.

File: torch/csrc/tensor/python_tensor.cpp [hunk 1]
```diff
@@ -434,27 +419,29 @@ static void py_bind_tensor_types(
...
+  // Checked whether passed object is in _tensor_classes set.
+  // Which is defined in torch/__init__.py, and get all supported items by
+  // call torch._tensor_classes in python script.
+  auto torch_tensor_classes = get_python_tensor_classes();
   PyTensorType* type = (PyTensorType*)obj;
   TORCH_CHECK_TYPE(
-      !type->is_cuda || torch::utils::cuda_enabled(),
-      "type ",
-      type->name,
-      " not available. Torch not compiled with CUDA enabled.")
```
Comment (ezyang) [inline torch/csrc/tensor/python_tensor.cpp:457]: So what happened to this test? It seems to have vanished?

File: torch/csrc/tensor/python_tensor.cpp [hunk 2]
```diff
@@ -434,27 +419,29 @@ static void py_bind_tensor_types(
...
 
-static bool PyTensorType_Check(PyObject* obj) {
-  auto it = std::find_if(
-      tensor_types.begin(), tensor_types.end(), [obj](PyTensorType* x) {
-        return (PyObject*)x == obj;
-      });
-  return it != tensor_types.end();
-}
-
 void py_set_default_tensor_type(PyObject* obj) {
```
Comment (ezyang) [inline torch/csrc/tensor/python_tensor.cpp:422]: Most of everything in this PR is non-objectionable, but the changes here seem a bit weird. What's going on?

Comment (Shan19900305) [reply torch/csrc/tensor/python_tensor.cpp:422]: PyTensorType_Check only checks object whether is in tensor_types, and it is assumed that the data stored on the torch._tensor_classes and tensor_types vector are consistent.
Now tensor_types are only used for CPU and GPU tensor types, which is a local static vector. But _tensor_classes is python side, which can be registered from other libraries for custom device tensor types. So from checked object based on local tensor_types vector to global torch._tensor_classes.

File: torch/csrc/tensor/python_tensor.cpp (hunk 1 above)
Comment (Shan19900305) [reply torch/csrc/tensor/python_tensor.cpp:457]: Oh, i overlooked this, add this back.

File: torch/csrc/tensor/python_tensor.cpp (hunk 2 above)
Comment (ezyang) [reply torch/csrc/tensor/python_tensor.cpp:422]: Drop this bit from this PR, if you really want it let's litigate it on its own. But this is the same "I don't want to aid and abet more usage of the legacy constructor style"

Comment (Shan19900305) [reply torch/csrc/tensor/python_tensor.cpp:422]: @ezyang ok, Dropped this bit in this PR, and i will create a issue for legacy constructor style.

**Review by ezyang — APPROVED** [review]

**Comment by Shan19900305** [comment]
> ## Merge failed
> **Reason**: 1 mandatory check(s) failed. The first few are:
> 
> * [pull / linux-jammy-py3.10-clang15-asan / test (default, 6, 6, linux.4xlarge)](https://github.com/pytorch/pytorch/actions/runs/8005574235/job/21868881543)
> 
> Dig deeper by [viewing the failures on hud](https://hud.pytorch.org/pytorch/pytorch/commit/39f3ffd30d1f5fac5d07c9bbce4d600d84247c02)
> 
> Details for Dev Infra team
> Raised by [workflow job](https://github.com/pytorch/pytorch/actions/runs/8013350811)
> Failing merge rule: Core Maintainers

Using new api name in torch/_dynamo/trace_rules.py file to fix this failure. @ezyang, Could you please take a look.

**Pull request closed without merging**